app/models/authentication.rb

```diff
@@ -142,10 +136,16 @@ def assign_values(options)
     self.attributes = options
   end
 
-  def self.build_credential_options
-    CREDENTIAL_TYPES.each_with_object({}) do |(k, v), hash|
-      hash[k] = v.constantize.descendants.each_with_object({}) do |klass, fields|
-        fields[klass.name] = klass::API_OPTIONS if defined? klass::API_OPTIONS
+  def self.credential_types
+    credential_subclasses = descendants.select { |d| d.try(:credential_type) }.sort_by(&:name)
+
+    credential_subclasses.each_with_object({}) do |klass, credential_hash|
+      unless credential_hash.key?(klass.credential_type.to_sym)
+        credential_hash[klass.credential_type.to_sym] = {}
+      end
```

Comment:
I think you can simplify to
```suggestion
      credential_hash[klass.credential_type.to_sym] ||= {}
```